hist/histv7/inc/ROOT/RHistImpl.hxx

```diff
@@ -227,122 +257,467 @@ namespace Internal {
+   void operator()(BINS & /*localBins*/, const AXES & /*axes*/, const COORD & /*coords*/) const
+   {}
 };
 
-/** Fill `coord` with low bin edge or center or high bin edge of all axes.
- */
-template <int I, int NDIM, class COORD, class AXES>
-struct RFillBinCoord {
-   void operator()(COORD &coord, const AXES &axes, EBinCoord kind, int binidx) const
-   {
-      constexpr const int thisAxis = NDIM - I - 1;
-      int axisbin = binidx % std::get<thisAxis>(axes).GetNBins();
-      switch (kind) {
-      case EBinCoord::kBinFrom: coord[thisAxis] = std::get<thisAxis>(axes).GetBinFrom(axisbin); break;
-      case EBinCoord::kBinCenter: coord[thisAxis] = std::get<thisAxis>(axes).GetBinCenter(axisbin); break;
-      case EBinCoord::kBinTo: coord[thisAxis] = std::get<thisAxis>(axes).GetBinTo(axisbin); break;
+template <int I, int NDIMS, typename BINS, typename COORD, class AXES>
+struct RCoordsToLocalBins {
+   void operator()(BINS &localBins, const AXES &axes, const COORD &coords) const
+   {
+      constexpr const int thisAxis = NDIMS - I - 1;
+      localBins[thisAxis] = std::get<thisAxis>(axes).FindBin(coords[thisAxis]);
+      RCoordsToLocalBins<I - 1, NDIMS, BINS, COORD, AXES>()(localBins, axes, coords);
+   }
+};
+
+/// Recursively converts local axis bins from the standard `kUnderflowBin`/`kOverflowBin` for
+/// under/overflow bin indexing convention, to the corresponding bin coordinates.
+template <int I, int NDIMS, typename BINS, typename COORD, class AXES>
+struct RLocalBinsToCoords;
+
+template <int NDIMS, typename BINS, typename COORD, class AXES>
+struct RLocalBinsToCoords<-1, NDIMS, BINS, COORD, AXES> {
+   void operator()(COORD & /*coords*/, const AXES & /*axes*/, const BINS & /*localBins*/, EBinCoord /*kind*/) const
+   {}
+};
+
+template <int I, int NDIMS, typename BINS, typename COORD, class AXES>
+struct RLocalBinsToCoords {
+   void operator()(COORD &coords, const AXES &axes, const BINS &localBins, EBinCoord kind) const
+   {
+      constexpr const int thisAxis = NDIMS - I - 1;
+      int axisbin = localBins[thisAxis];
+      if (axisbin == RAxisBase::kUnderflowBin) {
+         switch (kind) {
+            case EBinCoord::kBinFrom: coords[thisAxis] = std::get<thisAxis>(axes).GetBinFrom(axisbin); break;
+            case EBinCoord::kBinCenter: coords[thisAxis] = std::get<thisAxis>(axes).GetBinCenter(axisbin); break;
+            case EBinCoord::kBinTo: coords[thisAxis] = std::get<thisAxis>(axes).GetBinFrom(std::get<thisAxis>(axes).GetFirstBin()); break;
+         }
+      } else if (axisbin == RAxisBase::kOverflowBin) {
+         switch (kind) {
+            case EBinCoord::kBinFrom: coords[thisAxis] = std::get<thisAxis>(axes).GetBinTo(std::get<thisAxis>(axes).GetLastBin()); break;
+            case EBinCoord::kBinCenter: coords[thisAxis] = std::get<thisAxis>(axes).GetBinCenter(axisbin); break;
+            case EBinCoord::kBinTo: coords[thisAxis] = std::get<thisAxis>(axes).GetBinTo(axisbin); break;
+         }
```

Comment:
I think that with the new versions of RAxis `GetBin[From|Center|To]`, you do not need these special cases anymore, and can keep only the `else` version.